chain/client/src/debug.rs

```diff
@@ -382,87 +382,127 @@ impl ClientActor {
-                    0
-                },
-                gas_price_ratio: block.header().gas_price() as f64 / initial_gas_price as f64,
-            });
-            last_block_hash = block.header().prev_hash().clone();
-            last_block_timestamp = block.header().raw_timestamp();
-            last_block_height = block.header().height();
+                let chunks = match &block {
+                    Some(block) => block
+                        .chunks()
+                        .iter()
+                        .map(|chunk| DebugChunkStatus {
+                            shard_id: chunk.shard_id(),
+                            chunk_hash: chunk.chunk_hash(),
+                            chunk_producer: self
+                                .client
+                                .runtime_adapter
+                                .get_chunk_producer(
+                                    block_header.epoch_id(),
+                                    block_header.height(),
+                                    chunk.shard_id(),
+                                )
+                                .ok(),
+                            gas_used: chunk.gas_used(),
+                            processing_time_ms: CryptoHashTimer::get_timer_value(
+                                chunk.chunk_hash().0,
+                            )
+                            .map(|s| s.as_millis() as u64),
+                        })
+                        .collect(),
+                    None => vec![],
+                };
+
+                blocks.insert(
+                    block_hash,
+                    DebugBlockStatus {
+                        block_hash,
+                        prev_block_hash: *block_header.prev_hash(),
+                        block_height: block_header.height(),
+                        block_producer,
+                        is_block_missing: block.is_none(),
+                        is_on_canonical_chain,
+                        chunks,
+                        processing_time_ms: CryptoHashTimer::get_timer_value(block_hash)
+                            .map(|s| s.as_millis() as u64),
+                        block_timestamp: block_header.raw_timestamp(),
+                        gas_price_ratio: block_header.gas_price() as f64 / initial_gas_price as f64,
+                    },
+                );
+                // TODO(robin): using last epoch id when iterating in reverse height direction is
+                // not a good idea for calculating producer of missing heights. Revisit this.
+                last_epoch_id = block_header.epoch_id().clone();
+                if let Some(prev_height) = block_header.prev_height() {
+                    if block_header.height() != prev_height + 1 {
+                        // This block was produced using a Skip approval; make sure to fetch the
+                        // previous block even if it's very far back so we can better understand
+                        // the skip.
```

Comment:
Will this cause the page to be too long if there are many blocks being skipped in a row?

Reply:
Yea that's exactly why I don't like this heuristic as it is. But I don't yet have another heuristic that doesn't feel like an overkill and make this code look more complicated than it needs to be. If you have another idea let me know!

Reply:
What if we just show 50 blocks at a time, and add a button at the bottom of the page to expand the page to another 50 blocks if the user wants to see more previous blocks. If you think it's a good idea, you can do that in a separate PR as well.

Reply:
Yeah I want to do that, but I think that's a bit too much extra frontend logic that I wanted to leave it out to a separate PR. Also, if there are 1000 skips then it might take a while of scrolling before the parent shows up - that's the motivation I had for doing the extra querying.